### ISSUE TITLE
Renaming a result of file search fails

### ISSUE THREAD

**Issue by Andreas0602** [mucommander/mucommander#542]
Steps to reproduce the issue:

1. Go to your home directory
2. File > Find...
3. Search for filename *.log
4. In the list of search results, select any file
5. Press Shift + F6
6. Change the file name
7. Press Enter

Expected result: muCommander will rename the file
Actual result: Move error: Cannot write file

**Comment by ahadas** [comment]
Hi @Andreas0602 , as you seem to use file search, I'd like to get your opinion on this.
I see three options here:
1. To block operations on the results of the file search - users can still copy them, open them with viewer/editor, open/browse them, etc but won't be able to do things like renaming them, changing their date and so on. The downside is that it limits some functionality but on the other hand, it makes the search result immutable which simplifies things.
2. The other extreme is to enable also these operations that were blocked in the previous option and to change the search results accordingly. That means that if we search for all files that start with 'A' and we get `Arik`, then the user can rename it to `Arik2` and then the search result should include `Arik2` or the user can rename it to `Bar` and then it should drop from the search result. This would be better IMHO but doing that without re-triggering the search is pretty complex.
3. Some compromise between the two options above is to enable the aforementioned operations but not to update the search results. Let's consider the search I mentioned before, for searching all files that start with 'A' - if we rename `Arik` to `Arik2` then the search result will still show `Arik` but with an icon that shows that `Arik` doesn't exist. This is fairly simple to do without re-triggering the search although it can debatable when considering other fields like the file size (imagine that we look for all empty files and then the user opens a result with the text editor and add some content) or date (we don't yet enable search by modification date though)

I tend toward 3 but I see some sense in 1 as well
What do you think?

**Comment by Andreas0602** [comment]
I just checked how Apple handles this scenario in Finder. Here is what I did:

- In an empty directory, create two files: Anton.txt and Berta.txt
- Search for Anton --> Anton.txt will be shown and Berta.txt will not be shown
- Rename Anton.txt to Anton2.txt --> Anton2.txt will be shown
- Rename Anton2.txt to A_nton2.txt --> A_nton2.txt will disappear from the search results

So, Finder behaves according to option 2. And this is also what I would prefer personally.

I also tried to find out whether Finder will re-trigger the search when you rename a file in the search results, but I came to no clear conclusion. Finder uses Spotlight for searching, which makes it a bit difficult to find out what exactly happens behind the scenes. I only noticed one thing: When you change the name of a file with a third-party program, e.g. muCommander, while the file is listed in the search results in Finder, then Finder will show the modified name. So it seems like Finder permanently monitors and updates the search results --> Perfect!

Anyway, with option 2, re-triggering the search when the user renames a file (or performs other operations like changing the date) would be fine in my opinion.

If you don't want to implement option 2 for whatever reason, then then option 1 would be my second-favorite solution. Note that Finder also blocks some operations while search results are displayed -- for example, you cannot create a new folder.

Option 3 is not very intuitive for the user, in my opinion.

To summarize, please do it the Apple way, if possible. They know a lot about good user interface design.

Just my 2 cents.

**Comment by Andreas0602** [comment]
Here is the proof that Apple really monitors and updates the search results permanently:

- In an empty directory, create two files: Anton.txt and Berta.txt
- Search for Anton --> Anton.txt will be shown and Berta.txt will not be shown
- In a third-part program, for example muCommander, create the file Anton2.txt --> Finder will show the files Anton.txt and Anton2.txt in the search results

**Comment by ahadas** [comment]
Thanks for the quick feedback @Andreas0602 
Disabling parent folder operations on file search makes a lot of sense - pushed #569 for that.
And yeah, I agree that the second option would be best from the user experience point of view. On the one hand, I'm a bit reluctant to re-trigger the search because the search may take a long time but on the other hand, it is way simpler than keeping the search results in sync without re-triggering the search. So let's go with that and check the feedback we get to see if it is worth putting more efforts in improving this

**Comment by Andreas0602** [comment]
@ahadas, Testing this change is not possible right now, because the nightly build only contains source code, no binaries.

**Comment by ahadas** [comment]
Yeah, there's a sporadic failure in publishing the nightly release, I need to look into that
Anyway, I published it manually and it's now available

**Comment by Andreas0602** [comment]
Today's nightly build works fine. I just want to report one thing which seems strange. To reproduce it, follow these steps:

1. In an empty directory, create some files
2. Search with default settings --> All files will be found
3. Execute View > Set same folder

After step 2, the search results will not be periodically updated. As a consequence, changes done by a third-party program, e.g. renaming a file, will not be reflected in the search results. That's not nice.

However, after step 3, the search results in both panes will be periodically updated.

That's inconsistent.

One more thing: While search results are updated, the previous search results will disappear und the pane will be blank, until the new search results will be shown. This behavior is quite ugly and irritating from user's perspective. The behavior of Apples finder is different: Search results will always be shown. At no point in time the pane will be blank.

**Comment by ahadas** [comment]
@Andreas0602 thanks for the confirmation!
About that inconsistent behaviour - yeah, it looks like the search is re-triggered periodically when we check if the file table needs to be refreshed when the search is displayed in both panels. I guess both panels use the same "search file" and so each one updates the modification date in its turn - but I need to take a deeper look

**Comment by ahadas** [comment]
Merged #571 to fix that

**Comment by Andreas0602** [comment]
> Merged #571 to fix that

Good to hear. Unfortunately, there are no binaries available again.

**Comment by ahadas** [comment]
Published

**Comment by Andreas0602** [comment]
@ahadas, it works well now. Thank you very much!

I think the fact that muCommander does not recognize changes done by another program is not too big a problem. 

However, the fact that muCommander does not have a "go to file" function (see #543) is quite a significant limitation in my opinion. It would be great if you could implement it, in order to bring file search's functionality to the same level that other file managers have.

**Comment by ahadas** [comment]
Thanks for checking and for the feedback @Andreas0602 
Posted #572 - an upside of not caching the file searches is that it now works well when both panels show the same search and it might help with memory consumption but a downside is that the search is not preserved when switching between tabs so I open the new tab in the inactive panel. I'll get that in for the next nightly version and please let me know on #543 what you think

**Comment by ahadas** [comment]
Version 1.0.0 that includes a fix for this issue [has been released](https://github.com/mucommander/mucommander/releases/tag/1.0.0-1)
